bd-client-stats/src/stats.rs

```diff
@@ -485,47 +276,125 @@ impl<T: TimeProvider, F: SerializedFileSystem> Flusher<T, F> {
+    ) -> anyhow::Result<Option<oneshot::Receiver<UploadResponse>>> {
+      if let Some(pending_upload) = flusher
+        .file_manager
+        .get_or_create_pending_upload(only_if_file_is_old)
+        .await?
+      {
+        return flusher.process_pending_upload(pending_upload).await;
+      }
+      Ok(None)
+    }
+
+    // Note on error handling: while we could probably gracefully handle some of the failing I/O
+    // operations, it is likely to result in inaccurate stats (double submission of stats, missing
+    // aggregations, etc.), so we bail on failure. As we start seeing this out in the wild we may
+    // get a better understanding of why things are failing at which point we can do more targeted
+    // error handling.
+    log::debug!("processing upload from disk");
+    match inner(self, only_if_file_is_old).await {
+      Ok(result) => result,
+      Err(e) => {
+        handle_unexpected::<(), anyhow::Error>(Err(e), "upload from disk");
+        None
+      },
+    }
+  }
+
+  // Attempts to upload the provided stats request. Upon success, the file containing the pending
+  // request will be deleted.
+  async fn process_pending_upload(
+    &self,
+    request: StatsUploadRequest,
+  ) -> anyhow::Result<Option<oneshot::Receiver<UploadResponse>>> {
+    let (stats, response_rx) = TrackedStatsUploadRequest::new(request.upload_uuid.clone(), request);
+
+    log::debug!(
+      "sending pending flush upload: {} with {} metrics",
+      stats.payload.upload_uuid,
+      stats
+        .payload
+        .snapshot
+        .iter()
+        .map(|s| s.metrics().metric.len())
+        .sum::<usize>(),
+    );
+
+    let tracked_upload = DataUpload::StatsUploadRequest(stats);
+
+    // If this errors out the other end of the channel has closed, indicating that we are shutting
+    // down.
+    if self.data_flush_tx.send(tracked_upload).await.is_err() {
+      return Ok(None);
+    }
+
+    Ok(Some(response_rx))
+  }
+
+  async fn process_pending_upload_completion(
+    &self,
+    upload_response: UploadResponse,
+  ) -> Option<oneshot::Receiver<UploadResponse>> {
+    log::debug!("stat upload attempt complete: {:?}", upload_response);
+    if upload_response.success {
+      // If this fails we are in a bad state and are likely going to end up double uploading, but
+      // there is little we can do about it.
+      handle_unexpected(
+        self
+          .file_manager
+          .complete_pending_upload(&upload_response.uuid)
+          .await,
+        "complete pending upload",
+      );
+      // During startup or after getting network connectivity back it's possible that we will have
+      // a number of pending uploads to process. Go ahead and see if we have an old file at the
+      // head of the list which we should upload immediately.
+      // TODO(mattklein123): It would be better to batch all of the "old" files into a single
+      // upload request. We can do this in the future.
```

Comment:
If we are allocating a UUID per upload and use this for idempotency this might get tricky? Or would these pending uploads never have been attempted uploaded?

Reply:
The idea here would be to collapse the old files into a single upload and track it, but we have idempotency issues throughout the stack unfortunately, see all the TODOs in the SaaS. This is a really messy issue.